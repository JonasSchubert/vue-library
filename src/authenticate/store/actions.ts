import axios, { AxiosResponse } from 'axios';
import { ActionTree, Commit } from 'vuex';
import { ActionTypes, MutationTypes, RouteTypes } from './types';
import {
  AuthenticateState, LoginData, LoginResponse, ValidateTokenData
} from '../models';
import { snackbar } from '../../core/controls/app-snackbar';

const resetState = (commit: Commit): void => {
  commit(MutationTypes.setData, { data: undefined, saveLoginDataTemporary: false });
  commit(MutationTypes.setError, { error: undefined });
};

export const createActions = <TRootState, T extends LoginResponse>(): ActionTree<AuthenticateState<T>, TRootState> => ({
  [ActionTypes.login]({ commit, dispatch }, { password, saveLoginDataTemporary, userName }: LoginData): Promise<LoginResponse> {
    commit(MutationTypes.setIsLoading, { isLoading: true });
    resetState(commit);

    return new Promise((resolve) => {
      axios.get<LoginResponse>(`${RouteTypes.authenticate}/login/${userName}/${password}`)
        .then((response: AxiosResponse<LoginResponse>) => {
          if (response.data.success) {
            commit(MutationTypes.setData, { data: response.data, saveLoginDataTemporary });
          } else {
            snackbar.error({ message: 'message.login-failure', messageParams: [''] });
            const error: Error = {
              name: 'login-failure',
              message: 'message.login-failure'
            };
            commit(MutationTypes.setError, { error });
          }
        })
        .catch((error) => {
          snackbar.error({
            btn: {
              callback: () => dispatch(ActionTypes.login, { password, saveLoginDataTemporary, userName }),
              color: 'white',
              text: 'message.retry'
            },
            message: 'message.login-failure',
            messageParams: [error.message]
          });
          commit(MutationTypes.setError, { error });
        })
        .finally(() => {
          commit(MutationTypes.setIsLoading, { isLoading: false });
          resolve();
        });
    });
  },
  [ActionTypes.logout]({ commit, dispatch, getters }): Promise<boolean> {
    commit(MutationTypes.setIsLoading, { isLoading: true });

    return new Promise((resolve) => {
      axios.get<boolean>(`${RouteTypes.authenticate}/logout/${getters.token}`)
        .then((response: AxiosResponse<boolean>) => {
          if (response.data) {
            resetState(commit);
            snackbar.success({ message: 'message.logout-success' });
          } else {
            snackbar.error({ message: 'message.logout-failure' });
            const error: Error = {
              name: 'logout-failure',
              message: 'message.logout-failure'
            };
            commit(MutationTypes.setError, { error });
          }
        })
        .catch((error) => {
          snackbar.error({
            btn: {
              callback: () => dispatch(ActionTypes.logout),
              color: 'white',
              text: 'message.retry'
            },
            message: 'message.logout-failure',
            messageParams: [error.message]
          });
          commit(MutationTypes.setError, { error });
        })
        .finally(() => {
          commit(MutationTypes.setIsLoading, { isLoading: false });
          resolve();
        });
    });
  },
  [ActionTypes.validateToken]({ commit, dispatch }, { token }: ValidateTokenData): Promise<LoginResponse> {
    commit(MutationTypes.setIsLoading, { isLoading: true });
    resetState(commit);

    return new Promise((resolve) => {
      axios.get<LoginResponse>(`${RouteTypes.authenticate}/validate-token/${token}`)
        .then((response: AxiosResponse<LoginResponse>) => {
          if (response.data.success) {
            commit(MutationTypes.setData, { data: response.data, saveLoginDataTemporary: true });
          } else {
            snackbar.error({ message: 'message.token-validation-failure' });
            const error: Error = {
              name: 'token-validation-failure',
              message: 'message.token-validation-failure'
            };
            commit(MutationTypes.setError, { error });
          }
        })
        .catch((error) => {
          snackbar.error({
            btn: {
              callback: () => dispatch(ActionTypes.validateToken, { token }),
              color: 'white',
              text: 'message.retry'
            },
            message: 'message.token-validation-failure',
            messageParams: [error.message]
          });
          commit(MutationTypes.setError, { error });
        })
        .finally(() => {
          commit(MutationTypes.setIsLoading, { isLoading: false });
          resolve();
        });
    });
  }
});
