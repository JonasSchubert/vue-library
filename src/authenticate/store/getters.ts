import { GetterTree } from 'vuex';
import { GetterTypes } from './types';
import { UserRole } from '../enums';
import { AuthenticateState, LoginResponse } from '../models';

export const createGetters = <TRootState, T extends LoginResponse>(customGetters?: GetterTree<AuthenticateState<T>, TRootState>): GetterTree<AuthenticateState<T>, TRootState> => ({
  [GetterTypes.error]: (state: AuthenticateState<T>): Error | undefined => state.error,
  [GetterTypes.isAdministrator]: (state: AuthenticateState<T>): boolean => UserRole.toArray(state.data?.role ?? UserRole.Null).includes(UserRole.Administrator),
  [GetterTypes.isLoading]: (state: AuthenticateState<T>): boolean => state.isLoading,
  [GetterTypes.isUser]: (state: AuthenticateState<T>): boolean => UserRole.toArray(state.data?.role ?? UserRole.Null).includes(UserRole.User),
  [GetterTypes.loggedIn]: (state: AuthenticateState<T>): boolean => !!state.data?.token,
  [GetterTypes.token]: (state: AuthenticateState<T>): string | undefined => state.data?.token,
  [GetterTypes.userRoles]: (state: AuthenticateState<T>): UserRole[] => UserRole.toArray(state.data?.role ?? UserRole.Null),
  ...customGetters
});
