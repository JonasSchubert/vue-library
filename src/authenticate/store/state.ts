import { AuthenticateState, LoginResponse } from '../models';

export const createAuthenticateState = <T extends LoginResponse>(): AuthenticateState<T> => ({
  data: undefined,
  error: undefined,
  isLoading: false
});
