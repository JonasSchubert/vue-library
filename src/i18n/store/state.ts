import { I18nModuleConfig, I18nState } from '../models';

export const createState = ({ defaultLocale, defaultMessages }: I18nModuleConfig): I18nState => ({
  availableLocales: [defaultLocale],
  currentLocale: defaultLocale,
  error: undefined,
  locales: { [defaultLocale]: defaultMessages }
});
