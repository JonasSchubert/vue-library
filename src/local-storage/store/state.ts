import { LocalStorageState } from '../models';

export const createLocalStorageState = (isDarkTheme: boolean): LocalStorageState => ({
  history: {
    lastRoutes: []
  },
  settings: {
    isDarkTheme
  }
});
