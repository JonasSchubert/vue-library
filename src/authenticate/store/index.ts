import { Module } from 'vuex';
import { AuthenticateModuleSetup, AuthenticateState, LoginResponse } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createAuthenticateState } from './state';

export const createAuthenticateModule = <TRootState, T extends LoginResponse>({ cookieKeyAuthenticationToken, customGetters, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup<TRootState, T>): Module<AuthenticateState<T>, TRootState> => ({
  namespaced: true,
  actions: createActions(),
  getters: createGetters(customGetters),
  mutations: createMutations({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }),
  state: createAuthenticateState()
});
