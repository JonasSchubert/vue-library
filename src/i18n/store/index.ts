import { Module } from 'vuex';
import { I18nModuleConfig, I18nState } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createState } from './state';

export const createModule = <T>(config: I18nModuleConfig): Module<I18nState, T> => ({
  namespaced: true,
  actions: createActions<T>(config),
  getters: createGetters<T>(),
  mutations: createMutations(config),
  state: createState(config)
});
