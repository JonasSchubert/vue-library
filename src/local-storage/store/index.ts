import { Module } from 'vuex';
import { LocalStorageState } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createLocalStorageState } from './state';

export const createLocalStorageModule = <T>(vuetify: any): Module<LocalStorageState, T> => ({
  namespaced: true,
  actions: createActions<T>(vuetify),
  getters: createGetters<T>(),
  mutations: createMutations(),
  state: createLocalStorageState(vuetify.framework.theme.dark)
});
