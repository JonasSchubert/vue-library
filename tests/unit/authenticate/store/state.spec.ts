import { createAuthenticateState } from '@/authenticate/store/state';

describe('createAuthenticateState', () => {
  test('should not change unexpected', () => {
    // Arrange
    const authenticateState = createAuthenticateState();

    // Assert
    expect(authenticateState).toMatchSnapshot();
  });
});
