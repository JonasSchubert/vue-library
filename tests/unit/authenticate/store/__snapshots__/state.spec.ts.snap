// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAuthenticateState should not change unexpected 1`] = `
Object {
  "data": undefined,
  "error": undefined,
  "isLoading": false,
}
`;
